### PR TITLE
feat(cli): totem retrospect <pr> — bot-tax circuit-breaker (mmnto-ai/totem#1713)

### LINKED ISSUE #1713: feat(cli): bot-tax circuit breaker — totem retrospect <pr>
## Context

Surfaced from `mmnto-ai/liquid-city#97`'s 26-round bot-review marathon (3-5x prior maxes); held in `.journal/strategy/claude-0004-audit-fgh-prep-and-friction-reduction-arc.md:62-69` for Audit H grounding. Gemini-CLI cross-agent input on 2026-04-27 directed direct-file rather than continuing to wait on Audit H bandwidth (per `audits/internal/2026-04-27-strategic-intent-recovery-pre-fgh.md` § Resolution Q2).

## Diagnostic

*"The system has no step-back reflex."* When a PR exceeds N review rounds, the implicit protocol — every finding gets applied or defended in the open PR — is the **wrong default at high round counts**. R18-R26 of `mmnto-ai/liquid-city#97` was ~80% comment-drift findings; marginal value per round was low while marginal cost (quota + contributor + bot context-window) was constant.

Evidence: `upstream-feedback/032` (convergence-trigger discipline; status: new).

## Proposed shape

`totem retrospect <pr>` — invoked mid-PR when round count exceeds a threshold, forces a retrospective:

- Surfaces round-count + finding-distribution stats (per-vendor, per-class, dedup rate, override frequency)
- Identifies findings that should route to follow-up issues vs in-PR fixes
- Suggests stop-conditions ("if next round is purely comment-drift, ship + file followups")
- Optionally auto-files follow-up issues for findings the contributor confirms route-out

## Acceptance criteria

- [ ] CLI command exists at `packages/cli/src/commands/retrospect.ts` (or appropriate location)
- [ ] Invoking against a PR fetches review history and computes round-count + dedup metrics
- [ ] Output identifies candidates for "route to follow-up issue" vs "fix in PR"
- [ ] Threshold for invocation is configurable (per repo, sensible default like 5+ rounds)
- [ ] Documented surface (help text + example in `README.md` or workflow doc)

## Related

- `upstream-feedback/032` — convergence-trigger discipline (canonical exhibit: `mmnto-ai/liquid-city#97`)
- `upstream-feedback/033` — comment-drift heuristic for `totem review` (compositional with this feature)
- Memory: `feedback_product_vs_workflow_split.md` — this feature is a workflow-discipline graduating into product
- Memory: `feedback_four_honest_signals.md` — round-count is signal 1 + 3 (funnel shape + queue depth)

## Source

- `audits/internal/2026-04-27-strategic-intent-recovery-pre-fgh.md` § Tier 1.8
- `.journal/strategy/claude-0004-audit-fgh-prep-and-friction-reduction-arc.md:65`

### PR DESCRIPTION
**Pull request by satur8d**
## Mechanical Root Cause

PRs sometimes accumulate enough bot-review rounds that the marginal value-per-round collapses. `mmnto-ai/liquid-city#97` ran 26 rounds; rounds R18-R26 were ~80% comment-drift findings -- comments rotting behind code from earlier rounds, each fix triggering another full bot re-review wave (canonical exhibit per upstream-feedback `032-convergence-trigger-discipline-for-pr-review-cycles.md`). The contributor and the bots have no shared protocol for "this finding routes to follow-up issue, not in-PR fix." There's no command surface for the routing decision.

This is the third feature in the bot-tax cluster. `mmnto-ai/totem#1715` (`totem stats --pattern-recurrence`) shipped the substrate; `mmnto-ai/totem#1714` (`totem review --estimate`) shipped the pre-flight predictor. This PR ships the post-saturation circuit-breaker that reads the cross-PR substrate + live PR data and emits a deterministic route-out / in-PR-fix verdict per finding.

### `totem retrospect <pr-number>` (new top-level CLI command)

- **Deterministic, no-LLM.** Mirrors the `mmnto-ai/totem#1714` defense-in-depth pattern: static-source-grep on both new files (`packages/core/src/retrospect.ts` + `packages/cli/src/commands/retrospect.ts`) asserts no `from '@mmnto/totem-orchestrator'` / `getOrchestrator` / `Anthropic` / `OpenAI` / `gemini` import; runtime test mocks the orchestrator factory and asserts it is never called during `runRetrospect`.
- **Live PR fetch + substrate enrichment.** Uses `GitHubCliPrAdapter` for live review data; reads `.totem/recurrence-stats.json` (substrate, optional) for cross-PR-recurrence flags; reads `.totem/compiled-rules.json` (optional) for rule-coverage flags via Jaccard >= 0.6 against rule messages. Both substrate reads are graceful-degrade with `substrateAvailable` / `compiledRulesAvailable` boolean flags surfaced in the report.
- **Push-based round grouping.** Reviews are grouped by `commit_id` (head SHA at review time) -- one round per push, not one round per submission. New `GitHubCliPrAdapter.fetchReviews(prNumber)` method calls `gh api repos/.../pulls/N/reviews --paginate` because the existing `fetchPr` JSON shape doesn't carry `commit_id`.
- **Table-driven classifier.** `classifyFinding` is deterministic over the four-axis cube `(severityBucket x roundPosition x crossPrRecurrenceBucket x coveredByRule)`. `route-out` reasons come from a closed `RETROSPECT_ROUTE_OUT_REASONS` catalog so the report doesn't accumulate one-off prose strings. Stop-conditions are deterministic templates with substituted counts (no prose generation).
- **Sub-threshold skip is exit 0** (not exit 1) so the circuit-breaker doesn't fail CI on every invocation below threshold. `--force` overrides.
- **Optional `--out <path>` JSON write.** No file mutation by default; the only write surface is the explicit `--out` flag (deterministic two-space indent, mirrors `totem lint --format json --out`).

### Approved decisions (Q1-Q6 from `.totem/specs/1713.md`, confirmed by Gemini-via-relay)

- **Q1 -- round-grouping:** by `commit_id` (head SHA). Aligns with developer mental model (one push = one bot-review wave); robust to CR + GCA firing in quick succession on the same push.
- **Q2 -- `--auto-file` deferred:** v0.1 emits suggested route-out titles + bodies; user copy-pastes or pipes to `gh issue create`. Mass-filing is irreversible; convergence-trigger framing in upstream-feedback 032 emphasizes contributor-acknowledged routing. `// totem-context:` comment in `index.ts` documents the deferral pointing at a follow-up ticket.
- **Q3 -- threshold default:** 5. Same default as `totem stats --pattern-recurrence` for cluster consistency.
- **Q4 -- `--out <path>` JSON:** ships in v0.1. Enables `jq` and CI composition today.
- **Q5 -- comment-drift heuristic:** out of scope. Upstream-feedback 033 is a `totem review` feature, not retrospect. Cross-linked here.
- **Q6 -- postmerge ride-along:** branch off `postmerge/1727-1729-1732` so the curation commit (`9cb52c56`: 33 lessons + 5 compiled rules - 2 archived inline = 3 net new active rules) lands on this PR's first parent. Per Matt's bundle-locally directive.

### Spec deviation worth flagging (auto-spec gap, fourth recurrence)

The Gemini auto-spec recommended an LLM-driven classifier with a `--auto-file` flag that shells out to `gh issue create` for every route-out candidate. **Rejected** in the appended Implementation Design section -- it contradicts the design intent recorded in memory (deterministic slice of `recurrence-stats.json`) and the no-LLM-import pattern locked in by sibling `mmnto-ai/totem#1714`. Fourth recurrence of `feedback_auto_spec_gap` (after `mmnto-ai/totem#1665` / `mmnto-ai/totem#1688` / `mmnto-ai/totem#1690`); design doc supersedes auto-spec when prior-art patterns exist.

### Other principled deviations

1. **Severity-bucket re-export, not redefinition.** Extracted `toSeverityBucket` from `packages/cli/src/commands/recurrence-stats.ts` into a top-level export in `packages/core/src/recurrence-stats.ts`. Bot-tax cluster (`#1715` + `#1714` + `#1713`) now shares one severity vocabulary. Did not migrate the inline copy in the CLI's recurrence-stats command -- bytewise-identical, follow-up DRY pass can collapse without risk.
2. **`groupFindingsByRound` narrowed to bucket+order.** Inputs are `(submissions, findingsPerRoundCount)` rather than the brief's `(submissions, comments)`. The comments-to-submission timestamp join lives in the command (where the data crosses async boundaries) so the helper stays genuinely pure. Same end-to-end behavior; cleaner separation.
3. **`MinimalLogger.error` dropped.** First commit included it; project's ESLint id-match rule bans the literal identifier `error`. Per "always create new commits rather than amending," dropped in a follow-up commit (`7d6762ec`).
4. **Tenet 4 graceful-degrade catches documented.** Two catch blocks in `retrospect.ts` (substrate + compiled-rules read) take `// totem-context:` annotations matching the sibling `recurrence-stats.ts` pattern (commit `76b4b810`). Justified against the failure-mode table in `.totem/specs/1713.md` -- both downstream-data-absent paths surface `*Available: false` flags in the report so degradation is observable, not silent.

### Postmerge ride-along

This PR carries `9cb52c56` `chore: totem postmerge lessons for mmnto-ai/totem#1727 mmnto-ai/totem#1729 mmnto-ai/totem#1732` -- 33 lessons extracted, 5 compiled rules - 2 archived inline (40% bad rate, hits empirical baseline; one over-broad scope, one duplicate astGrepPattern). Net 3 new active rules. Plus auto-doc regen on `.github/copilot-instructions.md` + `.junie/skills/totem-rules/rules.md` (markdown escape flip-flop, `f2c08ad0`).

## Out of Scope

- LLM-mediated synthesis -- banned dependency in this command per the no-LLM invariant.
- GitHub mutation (`gh issue create`, `gh pr comment`, `gh pr review`) -- read-only.
- File mutation outside the optional `--out <path>` JSON write.
- Comment-drift detection (upstream-feedback 033 -- separate `totem review` ticket).
- Auto-batch follow-up issue filing (`--auto-file`) -- deferred per Q2.
- Trap-ledger writes -- read-only.
- Substrate freshness -- caller's responsibility to run `totem stats --pattern-recurrence` first.
- Round-position thresholds (`early <= 3` / `mid 4-9` / `late >= 10`) -- v0.1 magic numbers cleanly localized in `toRoundPosition` for future telemetry-driven calibration.
- Strategy submodule pointer not bumped per `feedback_no_auto_strategy_submodule_bump`.

## Tests Added/Updated

- [x] Added automated tests covering the root-cause mechanism

**`packages/core/src/retrospect.test.ts`** -- 82 new tests:

- Schema parse / reject for every Zod schema in the file.
- `it.each` over the full `(severityBucket, roundPosition, crossPrRecurrenceBucket, coveredByRule)` cube; every cell asserts classification + reason string from the closed catalog.
- Round grouping: head_sha collapsing, ordering by earliest `submitted_at` per SHA, missing-`commit_id` fallback bucket, single-round / multi-round / mixed-bot scenarios.
- Stop-condition templates (deterministic substitution of counts).
- Dedup math edge cases (zero findings, all unique, all dup).
- Signature parity with `recurrence-stats.ts` `computeSignature` / `normalizeFindingBody`.

**`packages/cli/src/commands/retrospect.test.ts`** -- 18 new tests:

- Sub-threshold skip exits 0; `--force` overrides.
- Substrate-missing graceful degrade: `substrateAvailable: false`; every finding `crossPrRecurrence: 0`.
- Substrate-malformed graceful degrade: same as missing, plus warning log.
- Compiled-rules-missing graceful degrade: `coveredByRule: false` for every finding.
- Zero bot comments edge case.
- `--out` write semantics: JSON shape, deterministic indent, no-write-by-default mtime snapshot.
- **No-LLM invariant, defense-in-depth:** static-source grep on both new files asserts no orchestrator / Anthropic / OpenAI / gemini imports. Runtime mock asserts orchestrator factory never called. Dynamic-import allowlist asserts every `await import(...)` argument resolves to a non-LLM module path.
- Cross-PR-recurrence target-PR exclusion: signature seen 5x on target PR + 0x elsewhere -> `crossPrRecurrence: 0`.
- Push-based round grouping over a constructed mock adapter response.
- Classification fan-out via `it.each`.

Test counts: **core 82/82 PASS** (285ms); **cli 18/18 PASS** (1.04s). Pre-push gates: `pnpm run format` (clean), `pnpm exec totem lint` (PASS, 0 errors, 181 pre-existing warnings), `pnpm exec totem verify-manifest` (PASS, 449 rules, hashes match).

Closes mmnto-ai/totem#1713

### CODE REVIEW

**Comment by satur8d** [comment]
@coderabbitai full review

**Comment by satur8d** [comment]
@coderabbitai full review

File: packages/cli/src/adapters/github-cli-pr.ts [hunk 1]
```diff
@@ -128,6 +152,35 @@ export class GitHubCliPrAdapter implements PrAdapter {
...
+      this.cwd,
+    );
+    return reviews.map((r) => ({
+      id: r.id,
+      user_login: r.user?.login ?? '',
+      commit_id: r.commit_id ?? undefined,
+      submitted_at: r.submitted_at ?? undefined,
+      state: r.state,
+      body: r.body ?? '',
+    }));
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/cli/src/adapters/github-cli-pr.ts:195]: Fixed in dfac2bca. `StandardPrReviewSubmission.user_login` widened to `string | null`; `fetchReviews` maps `r.user?.login ?? null`; null-guards added at both retrospect.ts call sites. Regression test added with `user_login: null` asserting the round count + total findings exclude it.

File: packages/cli/src/commands/retrospect.ts [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+  } else {
+    log.warn(
+      TAG,
+      `${rulesPath} not found — run 'totem lesson compile' first to enable rule-coverage enrichment.`,
+    );
+  }
+
+  // 6. Trap-ledger override events (read-only, count only).
+  const ledgerEvents = readLedgerEvents(totemDir, (msg) => log.warn(TAG, msg));
+  const overrideEventsObserved = ledgerEvents.filter((e) => e.type === 'override').length;
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/cli/src/commands/retrospect.ts]: Predicate scaffold applied as suggested in dfac2bca. The cast pattern is currently a no-op against the live LedgerEventSchema (no `prNumber` field today), so every event still passes — same global-count behavior as before, but forward-compat: when the schema gains `prNumber`, the filter starts scoping correctly with no command-side change. Filed mmnto-ai/totem#1736 to add the field at the ledger level.

File: packages/core/src/recurrence-stats.ts [hunk 3]
```diff
@@ -35,6 +35,47 @@ export const RecurrenceSeverityBucketSchema = z.enum(['critical', 'high', 'mediu
...
+  }
+  // unknown tool / synthesized review-body
+  if (s === 'critical') return 'critical';
+  // totem-context: `||` here is a chained equality fan-out across severity-label aliases, NOT a numeric default. The nullish-coalescing rule is a false positive on these comparison branches.
+  if (s === 'high' || s === 'major') return 'high';
+  // totem-context: chained equality fan-out — see comment above.
+  if (s === 'medium' || s === 'minor' || s === 'warning') return 'medium';
+  // totem-context: chained equality fan-out — see comment above.
+  if (s === 'low' || s === 'info') return 'low';
+  return 'nit';
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/recurrence-stats.ts:92]: Fixed in dfac2bca. Added explicit `if (tool === 'sarif')` branch with the SARIF v2.1.0 §3.27.10 `result.level` vocabulary (`error` → `high`, `warning` → `medium`, `note` / `none` → `low`). Test cube extended with 5 sarif rows; the type signature now matches the implementation.

**Comment by satur8d** [comment]
@coderabbitai round-1 summary — all findings addressed in `dfac2bca`.

**Inline (3 threads, replied individually):**
- `github-cli-pr.ts:181` user_login null — fixed (`string | null` + null-guards + regression test)
- `retrospect.ts:304` overrideEventsObserved scope — applied your cast scaffold; LedgerEventSchema doesn't carry `prNumber` today so the predicate is forward-compat. Filed mmnto-ai/totem#1736 to add the field at the ledger level.
- `recurrence-stats.ts:76` sarif handler — fixed (explicit branch + 5 test rows added).

**Review-body nitpicks:**
- Jaccard `0.6` extracted to `RULE_COVERAGE_JACCARD_THRESHOLD` (matches `DEFAULT_THRESHOLD` style).
- `.github/copilot-instructions.md:264+` duplicate directives — agreed this should be fixed at the exporter, not the diff. Filed mmnto-ai/totem#1735 for canonical-key + newest-timestamp dedup in the auto-doc export pipeline.

Tests: +5 core (sarif cube), +4 cli (3 input-validation, 1 null user_login regression). Lint PASS, 0 errors. Manifest verified (449 rules).

**Comment by satur8d** [comment]
@gemini-code-assist round-1 summary — all 3 high-priority findings addressed in `dfac2bca`.

1. **`retrospect.ts:121` — `parseInt` for prNumber.** Fixed: switched to `Number()` + `Number.isInteger()`; rejects `"5foo"`, `"5.2"`, leading/trailing whitespace, and non-positive values. Three new regression tests cover the rejection paths.

2. **`retrospect.ts:122` — `new Error('[Totem Error] ...')`.** Fixed: replaced with `TotemConfigError` (the canonical CLI-input class — `TotemConfigError` extends `TotemError`, and the `[Totem Error]` prefix is auto-prepended by the base constructor at `errors.ts:40`, so manual prefixing is removed). Recovery hint added: "Pass a positive integer (e.g. 'totem retrospect 1734')."

3. **`index.ts:459` — `parseInt` on `--threshold`.** Fixed: same `Number()` + `Number.isInteger()` pattern, with a `TotemConfigError` thrown for invalid values before the dynamic import of `runRetrospect`.

Lint PASS, 0 errors. Tests: 87 core / 22 cli on the retrospect surface, all green.

**Comment by satur8d** [comment]
@coderabbitai full review

**Comment by satur8d** [comment]
## Round-2 fixes (commits `da6e805d` + `b62ad97f`)

### CR actionable (6 → all fixed)

- **CR R2 (major)** `retrospect.ts:187` — replaced timestamp join with `pull_request_review_id` foreign-key join (your rebuttal was right; `created_at` predates `submitted_at` for pending reviews). Adapter schema gains `pull_request_review_id`; `StandardReviewComment` gains `pullRequestReviewId`. Regression test exercises the draft-review case explicitly.
- **CR R2 (major)** `retrospect.ts:577` — `sanitizeForTerminal` strips ANSI/control bytes before every `log.dim()` call site. Test confirms a hostile body (`\x1b[31mFAKE\x1b[2A\x07`) emerges sanitized.
- **CR R2 (minor)** `core/retrospect.ts:75` — `RetrospectFindingSchema` is now a Zod discriminated union on `classification`. `route-out` requires a `routeOutReason` from the closed `RETROSPECT_ROUTE_OUT_REASONS` catalog; `in-pr-fix` and `undetermined` forbid the field via `.strict()`. 5 schema tests cover each acceptance/rejection axis.
- **CR R2 (minor)** `core/retrospect.ts:367` — `buildStopConditions` denominator now includes `undetermined` findings. Regression test covers the false-fire case (1 covered route-out + 2 uncovered undetermined → 33% covered, no premature stop signal).
- **CR R2 (minor)** `lesson-bc2194f8.md` — wording requires random UUID suffix; PID/timestamp downgraded to optional extra entropy.
- **CR R2 (minor)** `cli-reference.md:223` — dropped "deferred to follow-up ticket" roadmap language; current-state framing only.

### CR nitpicks (5 → 3 fixed, 2 deferred)

- **CR-nit** `recurrence-stats.ts:50` — `toSeverityBucket` tool param now derives from `RecurrenceTool` (`Exclude<'mixed'>`).
- **CR-nit** `index.ts:430,464` — `RETROSPECT_DEFAULT_THRESHOLD` / `RETROSPECT_MIN_THRESHOLD` constants shared across help text + validation.
- **CR-nit** `retrospect.test.ts:511-538` — replaced source-only "runtime guard" with a real `vi.mock` of `../orchestrators/orchestrator.js` factories + `toHaveBeenCalledTimes(0)` assertion across a real `runRetrospect` run. You were right that the prior test was source inspection, not runtime.
- **CR-nit** `.github/copilot-instructions.md:407+` and `.junie/skills/totem-rules/rules.md:407+` — same auto-doc dedup as round-1's review-body finding. **Tracked in mmnto-ai/totem#1735** (export-pipeline canonical-key + newest-timestamp retention). Not in this PR.

### Pre-push Sonnet review (additional substrate fixes)

- **CRITICAL** missing `fetchReviews` unit tests → added (5 fixtures: bot, deleted-account, null-commit_id, PENDING, empty list).
- **WARN** `fetchReviews` was on the concrete adapter but not the `PrAdapter` interface → hoisted `StandardPrReviewSubmission` and `fetchReviews?` onto `pr-adapter.ts`. The interface now reflects what consumers can rely on.
- **INFO** `GhReviewCommentSchema.user` was non-nullable, would Zod-fail on inline comments from deleted accounts → made nullable + null-safe coercion in mapping (parallel to the round-1 user_login fix on review submissions).

### Test counts

| Suite | Before | After | Delta |
|---|---|---|---|
| `@mmnto/totem` | 1525 | 1531 | +6 |
| `@mmnto/cli` | 1900 | 1907 | +7 |

Lint PASS / 0 errors. Manifest verified (449 rules). Builds clean.

@coderabbitai full review
/gemini review

**Comment by satur8d** [comment]
@coderabbitai review
